### DISCUSSION TITLE
How to reset or spawnplayer inside game environment c++ code

### DISCUSSION THREAD

**Discussion by robispider** [microsoft/AirSim#3394] [Support Q&A]
Hello,

I am building a game environment with c++ code for my research project. the environment will be created from external commands. After building the environment I want to reset the player at a certain location. It has to be from environment c++ not from external c++ client. 
My steps:
1. I want to create an environment with my c++ code. (done) (but AirSimGame mode already spawned a vehicle at the beginning of the game and it fall to infinity as no ground plane yet to be created)
2. I want to place the vehicle at a certain location. (spawnplayer/  I can obtain the playerstart object and move it to my desired location. but can not spawn or reset the vehicle )
3. Start the game and after certain time I want to reset and goto step 2 (for resetting the vehicle to start point)
For this step 2:
Which c++ class from airsim plugin should I use?
WorldSimApi, ASimModeBase, AirBluePrintLib, or something else. 
I could not find any code comment to get idea of starting point. 
I am lost. Please help me with a starting point. A simple player reset/spawn  example with c++ will be greatly appreciated.

best regards, 

robi

**Comment by rajat2004** [comment]
Not entirely sure based on what you've described, so there could be multiple ways. First, is your requirement same as an existing API such as `spawnObject`? In that case, WorldSimApi would be the best place to start, you could just call the specific method of that class which implements the API e.g. [`spawnObject` in WorldSimApi](https://github.com/microsoft/AirSim/blob/master/Unreal/Plugins/AirSim/Source/WorldSimApi.cpp#L85)
ASimModeBase is and AirBlueprintLib are a bit more lower level, many methods in WorldSimApi will call those where required. Plus ASimModeBase has many startup stuff, etc which may or may not be what you need.

Since you mentioned creating a vehicle specifically, https://github.com/microsoft/AirSim/pull/2390 might be useful for you. It adds an API to create a vehicle at runtime. In that also, you could directly use the API implementaiton method in WorldSimApi

**Reply by robispider** [reply]
Hi,
Thank you very much for your response. 
I am actually having difficulty to figure out how to use spawnObject in WorldSimApi, as the function parameters are not commented. Can you please give a little code example. 

I also find spawnPlayer in [WorldSimApi ](https://github.com/microsoft/AirSim/blob/288c90ca78ce09ed58de8508622cb97071c8f217/Unreal/Plugins/AirSim/Source/WorldSimApi.cpp#L46)
but its a private function.  Can you tell what this function does?

best regards, 
robi

**Reply by robispider** [reply]
hi,
I also tried a code like this
                                `ASimModeBase* airgame=ASimModeBase::getSimMode();
				WorldSimApi::Pose pose;
				pose.zero();

				WorldSimApi::Vector3r scale;
				scale.Ones();

				WorldSimApi api = WorldSimApi(airgame);
				std::string str = "Drone1";
				api.spawnObject(str, "simpleFlight", pose, scale, true);`

it gave me a error like below:

> Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK1120	2 unresolved externals	maze_runner_v2	D:\zahid_practice\maze_runner v2\Binaries\Win64\UE4Editor-maze_runner-0003.dll	1	
Error	LNK2019	unresolved external symbol "public: __cdecl WorldSimApi::WorldSimApi(class ASimModeBase *)" (??0WorldSimApi@@QEAA@PEAVASimModeBase@@@Z) referenced in function "public: void __cdecl AMaze::SpawnTiles(void)" (?SpawnTiles@AMaze@@QEAAXXZ)	maze_runner_v2	D:\zahid_practice\maze_runner v2\Intermediate\ProjectFiles\Maze.cpp.obj	1	
Error	LNK2019	unresolved external symbol "public: virtual class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl WorldSimApi::spawnObject(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > &,class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const &,struct msr::airlib::VectorMathT<class Eigen::Matrix<float,3,1,0,3,1>,class Eigen::Quaternion<float,2>,float>::Pose const &,class Eigen::Matrix<float,3,1,0,3,1> const &,bool)" (?spawnObject@WorldSimApi@@UEAA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEAV23@AEBV23@AEBUPose@?$VectorMathT@V?$Matrix@M$02$00$0A@$02$00@Eigen@@V?$Quaternion@M$01@2@M@airlib@msr@@AEBV?$Matrix@M$02$00$0A@$02$00@Eigen@@_N@Z) referenced in function "public: void __cdecl AMaze::SpawnTiles(void)" (?SpawnTiles@AMaze@@QEAAXXZ)	maze_runner_v2	D:\zahid_practice\maze_runner v2\Intermediate\ProjectFiles\Maze.cpp.obj	1

**Reply by rajat2004** [reply]
Best place to understand the function parameters would be the Python API reference docs, for e.g. [`simSpawnObject`](https://microsoft.github.io/AirSim/api_docs/html/#airsim.client.VehicleClient.simSpawnObject)

FWIU, `spawnPlayer` seems to be used only for loading a level, to set the starting location in the new level. The naming might be a bit off, but doesn't seem to be useful for you case.

As to the C++ code, SimModeBase already has a [`world_sim_api_`](https://github.com/microsoft/AirSim/blob/288c90ca78ce09ed58de8508622cb97071c8f217/Unreal/Plugins/AirSim/Source/SimMode/SimModeBase.h#L160), which gets [initlialized in the beginning](https://github.com/microsoft/AirSim/blob/288c90ca78ce09ed58de8508622cb97071c8f217/Unreal/Plugins/AirSim/Source/SimMode/SimModeBase.cpp#L119). So creating a WorldSimApi onject doesn't seem to be the correct way to go about it. The references are somewhat of a mess when I think about it as well, since SimMode contains a pointer to WorlSimApi, and vice versa, and ApiProvder stores the pointer to WorldSimApi to call the methods in RpcLibServer. Simplest way I can think of is that you can add a getter method in SimModeBase, similar to [`getApiProvider`](https://github.com/microsoft/AirSim/blob/288c90ca78ce09ed58de8508622cb97071c8f217/Unreal/Plugins/AirSim/Source/SimMode/SimModeBase.h#L80) and use that to get the pointer and call the method as required.

The linking error is maybe due to some missing include or library settings which I can't say much about
